packages/serialize-request/lib/index.js

```diff
@@ -154,16 +154,33 @@ function serializeRequest(request, options = {}) {
  * Serialize request headers.
  *
  * @private
- * @param {Record<string, any>} headers
+ * @param {Headers | Record<string, any> | Iterable<[string, string]>} headers
  *     The headers object to serialize.
  * @param {Array<string>} includeHeaders
  *     An array of request headers to include.
  * @returns {SerializedRequestHeaders}
  *     Returns the serialized request headers.
  */
 function serializeHeaders(headers, includeHeaders) {
+	const headersObject = {};
+
+	if (Array.isArray(headers)) {
```

Comment:
Some clarification on this one, by type guarding I mean making sure that the headers we get are _actually_ strings. TypeScript will check this as a kind of linting step but there's nothing stopping someone passing in malformed data. It's safer if we do something like the following (feel free to reduce repetition / tweak obvs):

```js
if (Array.isArray(headers)) {
	for (const [header, value] of headers) {
		if (typeof header === 'string' && typeof value === 'string') {
			headersObject[header.toLowerCase()] = value;
		}
	}
} else if (isIterableHeadersObject(headers)) {
	for (const [header, value] of headers) {
		if (typeof header === 'string' && typeof value === 'string') {
			headersObject[header.toLowerCase()] = value;
		}
	}
} else {
	for (const header of Object.keys(headers)) {
		if (typeof header === 'string' && typeof value === 'string') {
			headersObject[header.toLowerCase()] = value;
		}
	}
}
```

Perhaps a less repetitive way of writing this would be something like:

```js
let iterableHeaders = [];

if (Array.isArray(headers) || isIterableHeadersObject(headers)) {
	iterableHeaders = headers;
} else {
	iterableHeaders = Object.keys(headers);
}

for (const [header, value] of iterableHeaders) {
	if (typeof header === 'string' && typeof value === 'string') {
		headersObject[header.toLowerCase()] = value;
	}
}
```